src/core/parse_worker.hpp

```diff
@@ -0,0 +1,238 @@
+ * proc() will execute the routine set by handler's passed function pointer;
+ * then, set to move on to the next handler. On the last handler,
+ * this will wrap around to the first, for so long as the passed "iters" limiter
+ * isn't exceeded (unlimited by default); once it is, this ParseWorker is said
+ * to be in a "done" state, on which proc() will no longer execute.
+ *
+ * Said to be in a "working" state when handling a TokenHandler after
+ * the first; this only matters to a TokenHandler which is holding
+ * this ParseWorker.
+ *
+ * Object of type dest_t is passed by reference to proc(), and is ultimately
+ * passed down to specified worker routines.
+ *
+ * TokenType must be compliant with the requirements of std::unordered_set
+ */
+
+template <class TokenType, class DestType>
+class ParseWorker 
+{
+	public:
+		using worker_t = ParseWorker;
+		using worker_arr_t = std::vector<worker_t>;
+		using worker_arr_init_t = std::initializer_list<worker_t>;
+
+		using token_t = TokenType;
+		using dest_t = DestType;
+
+		/*
+		 * A TokenHandler object holds an unordered set of symbols to match against (tokens_),
+		 * a function pointer which references a routine to execute on match (on_match_), and
+		 * an array of ParseWorker objects to process while still unmatched (workers_).
+		 *
+		 * If neg_ is set true, a given symbol will only match if it is not in tokens_.
+		 *
+		 * If working_ is not a nullptr, it indicates that one of the workers is currently in
+		 * the "working" state, and points to this worker; until it is done, only this worker
+		 * will be processed.
+		 */
+		class TokenHandler
+		{
+			friend class ParseWorker;
+			public:	
+				using token_set_t = std::unordered_set<token_t>;
+				using token_arr_init_t = std::initializer_list<token_t>;
+
+				using routine_t = typename WorkerRoutine<ParseWorker>::routine_t;
+
+				TokenHandler(token_arr_init_t t, routine_t on_m, worker_arr_init_t w={})
+					: tokens_(t), on_match_(on_m), workers_(w)
+				{}
+
+				TokenHandler(token_t t, routine_t on_m, worker_arr_init_t w={})
+					: tokens_{ t }, on_match_(on_m), workers_(w)
+				{}
+
+				TokenHandler(worker_arr_init_t w)
+					: TokenHandler(Symbol::END_OF_FILE, nullptr, std::move(w))
+				{}
+
+				bool match(const token_t& t);
+
+				TokenHandler& neg() { neg_ = true; return *this; }
+				void on_match(routine_t on_m) { on_match_ = on_m; }
+
+			private:
+				bool proc_workers_(const token_t& t, dest_t& f);
+				void reset_workers_();
+
+				token_set_t tokens_;
+				routine_t on_match_;
+				worker_arr_t workers_;
+
+				worker_t *working_ = nullptr;
+				bool neg_ = false;
+		};
+
+		using handler_t = TokenHandler;
+		using handler_arr_t = std::vector<handler_t>;
+		using handler_arr_init_t = std::initializer_list<handler_t>;
+
+		ParseWorker(handler_arr_init_t handlers)
+			: handlers_(std::move(handlers))
+		{}
+
+		explicit ParseWorker(handler_t handler)
```

Comment:
the more canoncial way is `const handler_t&` and no `std::move`

Reply:
If I do it this way, the compiler can construct `handler` from move constructor if it is passed an r-value. If you take `const handler_t&`, you are guarenteed a copy.

Reply:
If you really want this optimization, people usually have 2 constructors one with const handler_t& and handler&&. If you don't need pass by lvalue you onlyneed handler&&. This is good too because you know at compile time if you pass an lvalue cuz you'llget an error

Reply:
I like this person's response:
https://stackoverflow.com/questions/15673889/correct-usage-of-rvalue-references-as-parameters

Namely: 

> In general, if you take an argument by reference and copy it inside your function, then it's better to take the argument by value. In this case, instead of creating a temporary and copying it, the compiler can construct the argument

I've seen this sentiment in a few places—some think you should never really use r-value parameters outside of move constructor and assignment. Could you explain why this isn't the case here?

Reply:
Whoops—it appears I didn't read that post close enough. They do specifically note constructors (not just copy constructor) as an exception, in a case very similar to this. Will fix.